modal_utils/grpc_testing.py

```diff
@@ -76,45 +78,92 @@ async def intercepted_method(servicer_self, stream):
     return cls
 
 
+class ResponseNotConsumed(Exception):
+    def __init__(self, unconsumed_requests: List[str]):
+        self.unconsumed_requests = unconsumed_requests
+        request_count = Counter(unconsumed_requests)
+        super().__init__(f"Expected but did not receive the following requests: {request_count}")
+
+
 class InterceptionContext:
     def __init__(self):
         self.calls: List[Tuple[str, Any]] = []  # List[Tuple[method_name, message]]
-        self.custom_responses: dict[str, deque[List[Any]]] = defaultdict(deque)
+        self.custom_responses: Dict[str, List[Tuple[Callable[[Any], bool], List[Any]]]] = defaultdict(list)
 
     def add_recv(self, method_name: str, msg):
         self.calls.append((method_name, msg))
 
-    def add_response(self, method_name: str, custom_response: Optional[List[Any]] = None):
-        # adds one response to a queue of responses for method method_name
-        if custom_response is not None:
-            assert isinstance(
-                custom_response, list
-            ), "custom_response should be a list of messages sent back by the servicer"
-
-        if custom_response:
-            self.custom_responses[method_name].append(custom_response)
+    def add_response(
```

Comment:
Basically we add two new things to the `add_response` functionality:
1. A `request_filter` argument taking a filter function (in addition to the method name) so you can selectively respond when there are multiple requests to the same method
2. Automatic assertion at the end of the interception-block that all `add_response()` payloads have been consumed, so you don't accidentally pass a test even though the assigned value isn't even used